<a href="https://colab.research.google.com/github/geopard26/MDS/blob/main/Egorov_Pavel_final_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Probability
## Final project

In the final project you will develop your own classification algorithm based on probabilistic models.

In `data.zip`you can find a collection of texts in English, Italian, Spanish, German, French, Polish and Portuguese languages obtained from random Wikipedia articles, see e.g. `Spanish.txt`. (See corresponding `.source.txt` files and links therein for lists of authors.)

In [ ]:
with open("Spanish.txt") as f:
    print(f.read(500))

In [ ]:
with open("German.txt") as f:
    print(f.read(500))

This is our training data. These texts are preprocessed: only standard Latin characters kept, diacritics removed, punctuations removed, all letters converted to lowercase. We will use similar preprocessing for new texts that we have to classify. Here are some useful functions to do it.

### Preprocessing functions

In [233]:
import re
import unicodedata

### FROM: https://stackoverflow.com/a/518232/3025981
def strip_accents(s):
    return ''.join(c for c in unicodedata.normalize('NFD', s)
                   if unicodedata.category(c) != 'Mn')
### END FROM

def clean_text(s):
    return re.sub("[^a-z \n]", "", strip_accents(s))

### Learning: obtaining character frequencies

First of all, we have to find character relative frequencies in texts of each language. We will consider them as probability for character to appear in our multinomial model. Write function `get_freqs(text, relative)` that takes string `text` as input and returns dictionary which keys are all distinct characters occurred in `text` and values are frequencies (relative if `relative` is `True` and absolute otherwise). A similar function was discussed in Python videos previously. Note that Python treat `str` objects as a sequence of characters (e.g. if you try to iterate it).

In [234]:
def get_freqs(text, relative=False):
    ### BEGIN SOLUTION
    freqs = {}

    # Iterate over each character in the text
    for char in text:
        if char in freqs:
            freqs[char] += 1
        else:
            freqs[char] = 1

    if relative:
        total_chars = len(text)
        freqs = {char: count / total_chars for char, count in freqs.items()}

    return freqs
    ### END SOLUTION

In [235]:
assert get_freqs('Hello, World!') == {'H': 1, 'e': 1, 'l': 3, 'o': 2, ',': 1,
                                      ' ': 1, 'W': 1, 'r': 1, 'd': 1, '!': 1}

Now use function `get_freqs` to create a dictionary `lang_to_prob` which keys are names of languages (i.e. `'English'`, `'Italian'`, `'Spanish'`, `'German'`, `'French'`, `'Polish'`, `'Portuguese'`) and values are dictionaries of relative frequencies, obtained by processing of corresponding `.txt` files.

In [236]:
### BEGIN SOLUTION
def create_lang_to_prob(lang_files):
    lang_to_prob = {}
    for lang, file_path in lang_files.items():
        with open(file_path, 'r', encoding='utf-8') as file:
            text = file.read()
            freqs = get_freqs(text, relative=True)
            lang_to_prob[lang] = freqs
    return lang_to_prob

lang_files = {
    'English': 'English.txt',
    'Italian': 'Italian.txt',
    'French': 'French.txt',
    'German': 'German.txt',
    'Polish': 'Polish.txt',
    'Portuguese': 'Portuguese.txt',
    'Spanish': 'Spanish.txt'
}

lang_to_probs = create_lang_to_prob(lang_files)

### END SOLUTION

In [237]:
assert abs(lang_to_probs['Polish']['a'] - 0.08504245058355897) < 0.00001
assert abs(lang_to_probs['English']['x'] - 0.001387857977519179) < 1e-5
assert len(set(lang_to_probs['Portuguese'])) == 28

### Likelihoods
Now let us start implementing the actual classifier. We begin with multinomial likelihood function. Implement function `multinomial_likelihood(probs, freqs)` that takes two arguments: `probs` are dictionary of probabilities of each character (in some language) and `freqs` is dictionary of absolute frequencies of each character (in some text we want to classify). This function have to return probability to obtain these absolute frequencies from multinomial distribution with given probabilities $P((X_1 = f_1) \cap (X_2 = f_2) \cap \ldots \cap (X_k = f_k))$ provided that $(X_1, \ldots, X_k)$ is a system of multinomially distributed values with probabilities $(p_1, \ldots, p_k)$. (You need `factorial` function that can be imported from `math`.)

In [238]:
### BEGIN SOLUTION
from math import factorial

def multinomial_likelihood(probs, freqs):
    n = sum(freqs.values())  # Total number of observations
    likelihood_coefficient = factorial(n)

    for char, freq in freqs.items():
        likelihood_coefficient /= factorial(freq)
        likelihood_coefficient *= probs.get(char, 0) ** freq

    return likelihood_coefficient
### END SOLUTION

Let's find likelihood of data with frequencies `{'a': 2, 'b': 1, 'c': 2}` and probabilities `{'a': 0.2, 'b': 0.5, 'c': 0.3}`:

In [239]:
multinomial_likelihood(probs={'a': 0.2, 'b': 0.5, 'c': 0.3}, freqs={'a': 2, 'b': 1, 'c': 2})

0.054000000000000006

In [240]:
assert abs(multinomial_likelihood(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                    freqs={'a': 2, 'b': 1, 'c': 2}) - 0.054) < 0.000001
assert abs(multinomial_likelihood(probs={'a': 0.2, 'b': 0.1, 'c': 0.3, 'd': 0.4},
                    freqs={'a': 2, 'b': 1, 'c': 2}) - 0.0108) < 0.000001

Note that the coefficient with factorials depends only on `freqs` (i.e. text that we analyse) and does not depend on `probs`. It means that for all possible languages this coefficient will be the same. As we are going to consider fixed text and compare likelihoods for different languages, we see, that we do not need this coefficient in most of cases. Let us implement function `multinomial_likelihood_without_coeff` that returns the same probability as `multinomial_likelihood` but without coefficient.

In [243]:
### BEGIN SOLUTION
from numpy import power

def multinomial_likelihood_without_coeff(probs, freqs):
    p = 1
    for c in freqs:
        p *= power(probs[c], freqs[c])
    return p
### END SOLUTION

Corresponding probability become smaller:

In [244]:
multinomial_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                                     freqs={'a': 2, 'b': 1, 'c': 2})

0.0018000000000000004

In [245]:
assert multinomial_likelihood_without_coeff(
    probs={'a': 0.3, 'b': 0.4, 'c': 0.3},
    freqs={'a': 2, 'b': 1, 'c': 2}) == 0.00324
assert abs(multinomial_likelihood_without_coeff(
    probs={'a': 0.3, 'b': 0.4, 'c': 0.3},
    freqs={'a': 2, 'b': 1, 'c': 5}) - 8.747999999999e-05) < 1e-10

Actually, likelihoods become extremely small very quickly when we increase absolute frequencies in data. It is not surprisingly: the probability to get the text that coincides with our actual text from random experiment we discussed is extremely small.

In [246]:
multinomial_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                    freqs={'a': 3, 'b': 2, 'c': 2})

0.00018000000000000004

In [247]:
multinomial_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                    freqs={'a': 3, 'b': 20, 'c': 2})

6.866455078125001e-10

Due to limited precision of computer arithmetic, for frequencies large enough we will get exactly zero likelihood.

In [248]:
multinomial_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                                     freqs={'a': 543, 'b': 512, 'c': 2})

0.0

Thus we usually cannot use likelihoods like this directly. Common way to deal with such a tiny numbers is to use _logarithms_ instead of likelihood themself. Indeed, logarithm is monotonically increasing function. If we compare logarithms, it is equivalent to compare their arguments.

### Log likelihood
Implement function `log_likelihood_without_coeff(probs, freqs)` that calculates logarithm of likelihood (without factorial coefficient). Note that you cannot simply put `multinomial_likelihood_without_coeff` into `log`: if the likelihood would be extremely small (equal to zero from computer's point of view), `log` of it will be negative infinity. Thus we have to algebraically transform the expression for log likelihood first. Let us use properties of logarithm to do it: $\log (ab) = \log a + \log b$, $\log (a^b)=b\log a$.

In [249]:
### BEGIN SOLUTION
from math import log

def log_likelihood_without_coeff(probs, freqs):
    p = 0
    for c in freqs:
        p += freqs[c] * log(probs[c])
    return p

### END SOLUTION

Likelihoods are probabilities, so they are less than 1 and their logarithms are negative. The larger absolute value of log-likelihood, the less is likelihood.

In [250]:
log_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                             freqs={'a': 2, 'b': 1, 'c': 2})

-6.319968614080018

Now we can deal with inputs that lead to exact zero value previously.

In [251]:
log_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                             freqs={'a': 543, 'b': 512, 'c': 2})

-1231.2240885070605

In [252]:
assert abs(log_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                              freqs={'a': 2, 'b': 1, 'c': 2}) + 6.319968614080018) < 0.00001
assert abs(log_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                             freqs={'a': 543, 'b': 512, 'c': 2}) + 1231.2240885070605) < 0.0001
assert abs(log_likelihood_without_coeff(probs={'a': 0.2, 'b': 0.5, 'c': 0.3},
                             freqs={'a': 543, 'b': 512}) + 1228.8161428984085) < 0.0001

### Theoretical question: cross-entropy and Jensen's inequality

The function that you obtained is also know as *cross entropy* and is extremely popular in machine learning, namely, in classification problems. It allows you to measure how good probability distribution $(p_1, \ldots, p_k)$ fits to actual absolute frequencies obtained from the data $(f_1, \ldots, f_k)$.

Assume that frequencies $(f_1, \ldots, f_k)$ are fixed. What is the best distribution $(p_1, \ldots, p_k)$ from likelihood's perspective?

Intuitively, it seems that we have to put relative frequencies
$$r_i = \frac{f_i}{\sum_{j=1}^k f_j}$$
as $p_i$ to get best fit. In fact, it is true. To prove it, let us use Jensen's inequality for logarithms. It is stated as follows:

For any values $\alpha_1, \ldots, \alpha_k$, such that $\sum_{j=1}^k \alpha_j = 1$ and $\alpha_j \ge 0$ for all $j=1, \ldots, k$, and any positive values $x_1, \ldots, x_k$, the following inequality holds:

$$\log \sum_{j=1}^k \alpha_j x_j \ge \sum_{j=1}^k \alpha_j\log(x_j).$$

Use this inequality to prove that
$$\sum_{j=1}^k r_j \log p_j - \sum_{j=1}^k r_j \log r_j \le 0,$$
then prove that to obtain maximum log-likelihood (and therefore maximum likelihood) for fixed $(f_1, \ldots, f_k)$ we have to put $p_i=r_i$, $i=1,\ldots, k$.

**Hint:** use properties of logarithm to transform left-hand part of the last inequality to right-rand part of the previous inequality.

(Submit your proof as staff graded assignment.)

## Solution

### Proof of Using Relative Frequencies to Maximize Log-Likelihood

To prove that the best probability distribution $(p_1, \ldots, p_k)$ corresponds to the relative frequencies $(r_1, \ldots, r_k)$, where $r_i = \frac{f_i}{\sum_{j=1}^k f_j}$, we will use Jensen's Inequality for concave functions, such as the logarithmic function.

**Application of Jensen's Inequality**:

Jensen's Inequality states that for any numbers $\alpha_1, \ldots, \alpha_k$, where $\sum_{j=1}^k \alpha_j = 1$ and $\alpha_j \ge 0$, and any positive numbers $x_1, \ldots, x_k$, the following inequality holds:

$$\log \sum_{j=1}^k \alpha_j x_j \ge \sum_{j=1}^k \alpha_j\log(x_j).$$

In our case, let $\alpha_j = r_j$ and $x_j = \frac{p_j}{r_j}$. Applying Jensen's Inequality, we get:

$$ \log \sum_{j=1}^k r_j \cdot \frac{p_j}{r_j} \ge \sum_{j=1}^k r_j \log \frac{p_j}{r_j} $$

Since $\sum_{j=1}^k r_j = 1$ and $\sum_{j=1}^k p_j = 1$, the left side of the inequality simplifies to $\log(1) = 0$.

**Choosing $p_j = r_j$ to Maximize Log-Likelihood**:

Thus, $\sum_{j=1}^k r_j \log \frac{p_j}{r_j} \le 0$. Equality is achieved if and only if $\frac{p_j}{r_j} = 1$ for all $j$, i.e., when $p_j = r_j$ for all $j$. This condition corresponds to maximum log-likelihood because otherwise, the log-likelihood would be less than zero.

Therefore to maximize the log-likelihood for fixed frequencies $(f_1, \ldots, f_k)$, it is necessary to choose $p_j = r_j$ for all $j$.


### Maximum likelihood classifier

Now we will use likelihood to choose the best language for some text we want to classify. Write function `mle_best(text, lang_to_probs)` that takes some `text` and dictionary `lang_to_probs` that we created previously and returns the name of language such that likelihood of our data for this language is maximal. Note that you have to preprocess  `text` using function `clean_text` before finding frequencies.

In [253]:
### BEGIN SOLUTION
def mle_best(text, lang_to_probs):
    # Preprocess the text
    cleaned_text = clean_text(text)

    # Calculate frequencies of characters in the cleaned text
    freqs = get_freqs(cleaned_text)

    # Initialize variables to store the best language and its log likelihood
    best_lang = None
    # best_log_likelihood = float('-inf')
    highest_p = 0

    for lang in lang_to_probs:
        log_likelihood = multinomial_likelihood_without_coeff(lang_to_probs[lang], freqs)
        if log_likelihood > highest_p:
            best_lang = lang
            highest_p = log_likelihood
    return best_lang
### END SOLUTION

Let's test it!

In [254]:
# Source: https://en.wikipedia.org/wiki/1134_Kepler
text_en = """1134 Kepler, provisional designation 1929 SA, is a stony asteroid
and eccentric Mars-crosser from the asteroid belt, approximately
4 kilometers in diameter"""
mle_best(text_en, lang_to_probs)

'English'

In [255]:
# Source: https://pl.wikipedia.org/wiki/(1134)_Kepler
text = """"(1134) Kepler – planetoida z grupy przecinających
orbitę Marsa okrążająca Słońce w ciągu 4 lat i 145 dni
w średniej odległości 2,68 au.
"""
mle_best(text, lang_to_probs)

'Polish'

In [256]:
# Source: https://it.wikipedia.org/wiki/1134_Kepler
text = """1134 Kepler è un asteroide areosecante. Scoperto nel 1929,
presenta un'orbita caratterizzata da un semiasse maggiore pari a 2,6829098
UA e da un'eccentricità di 0,4651458, inclinata di 15,17381° rispetto
"""
mle_best(text, lang_to_probs)

'Italian'

In [257]:
lines = ['Aaa', 'Aa', 'Kepler è un asteroide areosecante. Scoperto nel',
        "presenta un'orbita caratterizzata da un semiasse maggiore pari a 2,6829098 "
         "UA e da un'eccentricità di 0,4651458, inclinata",
         "Kepler – planetoida z grupy przecinających orbitę Marsa okrążająca Słońce w ciągu 4 lat i ",
         "Kepler, provisional designation 1929 SA, is a stony asteroid "
         "and eccentric Mars-crosser from the"]
assert mle_best(lines[0], lang_to_probs) == 'Portuguese'
assert mle_best(lines[1], lang_to_probs) == 'Portuguese'
assert mle_best(lines[2], lang_to_probs) == 'French'
assert mle_best(lines[3], lang_to_probs) == 'Italian'
assert mle_best(lines[4], lang_to_probs) == 'Polish'
assert mle_best(lines[5], lang_to_probs) == 'English'
assert mle_best("Aaaa", lang_to_probs) == "Portuguese"

### Look, it works!
Fantastic! Just a couple of equations, and we created fully automatic language detection algorithm!

Now let us make it even better.

### Let's go Bayes
Assume that we selected random article from all Wikipedia articles in languages that we consider. There are different number of articles in different languages, so it is more likely to obtain article e.g. in English than in Polish (at least, at this moment). It means that we need stronger evidence for Polish to accept it comparing with English. To take into account this information, we will use Bayes' rule as discussed in the videos.

Recall that in Bayessian approach we consider _prior_ probabilities of languages, then use Bayes' rule to find their posterior probabilies and select language with largest posterior probability. We begin with finding prior probabilities. Create a dictionary `lang_to_prior` which keys are language names and values are prior probabilities. Assume that priors are prortional to number of articles in each language (in thousands). Let us use these numbers (in thousands): English: 6090, Italian: 1611, Spanish: 1602, German: 2439, French: 2222, Polish: 1412, Portuguese: 1034. Remember that all prior probabilities should sum up to 1!

In [259]:
### BEGIN SOLUTION
# Numbers of articles for each language (in thousands)
articles = {
    'English': 6090,
    'Italian': 1611,
    'Spanish': 1602,
    'German': 2439,
    'French': 2222,
    'Polish': 1412,
    'Portuguese': 1034
}

# Calculate the total number of articles
total_articles = sum(articles.values())

# Calculate prior probabilities
lang_to_prior = {lang: count / total_articles for lang, count in articles.items()}
### END SOLUTION

In [260]:
assert abs(lang_to_prior['French'] - 0.13540524070688603) < 0.000001

Now implement function `bayesian_best(text, lang_to_probs, lang_to_prior)` that takes some text `text`, dictionary `lang_to_probs` created before and `lang_to_prior` with prior probabilities. This function have to return language name with largest posterior probability. Note that as we only compare posterior probabilities, we can ignore the denominator in Bayes' rule: it is the same for all languages. Note also that we need to work with logarithms of posterior instead of posteriors themself to avoid extremely small numbers. Use properties of logarithm to do it effeciently.

In [261]:
### BEGIN SOLUTION
import math

def bayesian_best(text, lang_to_probs, lang_to_prior):
    text = clean_text(text)
    freqs = get_freqs(text)
    best_lang = None
    highest_p = 0
    for lang in lang_to_probs:
        probability_test = multinomial_likelihood_without_coeff(lang_to_probs[lang], freqs)
        probability_test *= lang_to_prior[lang]
        if probability_test > highest_p:
            best_lang = lang
            highest_p = probability_test
    return best_lang
### END SOLUTION

For example, MLE algorithm believes that word `"The"` belongs go German language.

In [262]:
mle_best("The", lang_to_probs)

'German'

However, if we take into account that English is more popular in Wikipdia, results changes:

In [263]:
bayesian_best("The", lang_to_probs, lang_to_prior)

'English'

In [264]:
bayesian_best("Kepler è un asteroide areosecante. Scoperto nel", lang_to_probs, lang_to_prior)

'French'

In [265]:
lines = ['Aaa', 'Aa', 'Kepler è un asteroide areosecante. Scoperto nel',
        "presenta un'orbita caratterizzata da un semiasse maggiore pari a 2,6829098 "
         "UA e da un'eccentricità di 0,4651458, inclinata",
         "Kepler – planetoida z grupy przecinających orbitę Marsa okrążająca Słońce w ciągu 4 lat i ",
         "Kepler, provisional designation 1929 SA, is a stony asteroid "
         "and eccentric Mars-crosser from the"]
answers = ['English', 'English', 'French', 'Italian', 'Polish', 'English']
for line, answer in zip(lines, answers):
    assert bayesian_best(line, lang_to_probs, lang_to_prior) == answer

### Measuring uncertainty
Finally, let us get posterior probability how certain our Bayesian algorithm in its classification. Implement function `bayesian_posterior(text, lang_to_probs, lang_to_prior, test_lang)` that takes some `text`, `lang_to_probs`, `lang_to_prior` and a particular language `test_lang` we are interested in and returns posterior probability for this language provided this text. Use formula for the posterior from the videos.

Note that in this case you have to work with actual probabilites (likelihoods), not their logarithms.

In [266]:
### BEGIN SOLUTIONS
def bayesian_posterior(text, lang_to_probs, lang_to_prior, test_lang):
  # Preprocess the text and calculate relative frequencies
    text = clean_text(text)
    freqs = get_freqs(text)
   # Calculate the posterior probability for the test language
    test_lang_posterior = multinomial_likelihood_without_coeff(lang_to_probs[test_lang], freqs)
    test_lang_posterior *= lang_to_prior[test_lang]
    total_posterior = 0
  # Calculate the sum of posterior probabilities for all languages
    for lang in lang_to_probs:
        lang_likelihood = multinomial_likelihood_without_coeff(lang_to_probs[lang], freqs)
        lang_likelihood *= lang_to_prior[lang]
        total_posterior += lang_likelihood
    return test_lang_posterior / total_posterior
### END SOLUTIONS

In [267]:
bayesian_posterior("The", lang_to_probs, lang_to_prior, "German")

0.2686381464045339

In [268]:
bayesian_posterior("The", lang_to_probs, lang_to_prior, "English")

0.534626604689872

In [269]:
bayesian_posterior("Das", lang_to_probs, lang_to_prior, "English")

0.36596793151737517

In [270]:
bayesian_posterior("Das", lang_to_probs, lang_to_prior, "German")

0.12095519926831602

In [271]:
assert abs(bayesian_posterior("The", lang_to_probs, lang_to_prior, "German") - 0.26863814640) < 0.00001
assert abs(bayesian_posterior("The", lang_to_probs, lang_to_prior, "English") - 0.534626604) < 0.00001
assert abs(bayesian_posterior("Das", lang_to_probs, lang_to_prior, "English") - 0.365967931517) < 0.00001

We see that our algorithm believes that `"Das"` belongs to English. This is due to small amount of data (only three letters!) and high prior for English. However, it is not very certain: the posterior is only 0.37. On the other hand, `"The"` belongs to `"English"` with much larger posterior probability.

## Conclusions
Today we used our knowledge of probability to construct our own classifier algorithm. Actually, it is a version of well-known naive Bayesian classifier. Of course, this classifier is far from perfect: for example, it completely ignores words and deal only with characters and their frequencies. Nevertheless, it works rather good despite being so simple. Also it shows several important concepts: likelihood, maximum likelihood estimation, Bayesian estimations and so on.

Now you are ready for more sophisticated topics.

### P.S. Efficiency considerations
During this project we used dictionaries to represent frequency tables and pure Python constructions like loops to process them. However, it is much more effecient to use `numpy` arrays and vectorized functions. You can try to redo this project with `numpy` if you know how to use it. If you don't know yet, that's not a problem: you will learn it soon.